feat: Filter unshared tasks out of clone list and task library [SCHOOL-202, SCHOOL-208]

BillClinton opened the pull request:
Closes: SCHOOL-202

The clone menu and task library both currently show all tasks regardless of whether they had their “shared” option checked or not.

Filter non-shared tasks out of both the clone menu and task library

Add fixture data as needed so we have a set of shared and non-shared tasks with corresponding titles, and extend existing automated tests for the teacher tasks dashboard and task library to verify that the clone task menu and the task library both exclude unshared tasks and include shared ones.

Bonus: add an “include unshared” option to the task library’s advanced menu. It is fine for any teacher with access to the task library to be able to browse unshared tasks.

BillClinton committed:
feat: add shared option and title to settings menu

BillClinton committed:
feat: correct minor layout issues on settings menu

BillClinton committed:
feat: add handler for shared option in settings

BillClinton committed:
feat: add condition to filter by shared status

BillClinton committed:
fix: correct name of option: shared to unshared

BillClinton committed:
fix: update test to work with new option feature

themightychris commented on php-classes/Slate/CBL/Tasks/TasksRequestHandler.php:
```diff
@@ -190,4 +197,11 @@ public static function getArchivedConditions($includeArchived) {
         }
     }
+
+    public static function getSharedConditions($includeShared) {
+        $recordClass = static::$recordClass;
+        if (!$includeShared) {
+            return $recordClass::getTableAlias().'.Status != "shared"';
+        }
+    }
```
```suggestion
```

BillClinton replied:
this function is no longer in use and has been removed

themightychris commented on php-classes/Slate/CBL/Tasks/TasksRequestHandler.php:
```diff
@@ -49,6 +49,13 @@ protected static function buildBrowseConditions(array $conditions = array(), arr
                 $conditions[] = $archivedConditions;
             }
+
+            $includeUnshared = isset($_REQUEST['include_unshared']) && ($_REQUEST['include_unshared']=='true');
+            $sharedConditions = static::getSharedConditions($includeUnshared);
+
+            if ($sharedConditions != false) {
+                $conditions[] = $sharedConditions;
+            }
```
```suggestion
            $includeUnshared = isset($_REQUEST['include_unshared']) && ($_REQUEST['include_unshared']=='true');
            $sharedConditions = static::getSharedConditions($includeUnshared);

            if (!$includeUnshared) {
                $conditions['Status'] = 'shared';
            }
```

BillClinton replied:
this has been resolved.  line #58 has also been removed as the $sharedConditions variable is no longer being used.

BillClinton committed:
fix: correct filter per internal sync feedback

BillClinton committed:
fix: remove unnecessary filter action

BillClinton committed:
feat: disable shared checkbox for new tasks

BillClinton committed:
feat: increase clarity/specificity of ref query

BillClinton committed:
fix: re-enable shared checkbox in case of disabled

BillClinton committed:
feat: add fixture data for testing shared filters

BillClinton committed:
feat: add test for shared/unshared grid filtering

BillClinton committed:
feat: add test for shared filter in clone select

BillClinton committed:
feat: clone select test to teacher tasks ui

BillClinton committed:
fix: remove commented line, var no longer in use

themightychris commented on sencha-workspace/SlateTasksManager/app/controller/Tasks.js:
```diff
@@ -155,6 +159,13 @@ Ext.define('SlateTasksManager.controller.Tasks', {
+    onIncludeUnsharedCheckboxClick: function(checkbox) {
+        var tasksStore = this.getTasksStore();
+
+        tasksStore.getProxy().extraParams.include_unshared = checkbox.checked;
```
Better to use the API method:

```suggestion
        tasksStore.getProxy().setExtraParam('include_unshared', checkbox.checked);
```

BillClinton replied:
resolved: [9bd19ea](https://github.com/SlateFoundation/slate-cbl/pull/710/commits/9bd19ea543cca4730c594bcee0f294406ffa6b6c)

themightychris commented on sencha-workspace/SlateTasksManager/app/controller/Tasks.js:
```diff
@@ -155,6 +159,13 @@ Ext.define('SlateTasksManager.controller.Tasks', {
+        var tasksStore = this.getTasksStore();
+
+        tasksStore.getProxy().extraParams.include_unshared = checkbox.checked;
+        tasksStore.load();
```
Looks like if you're on the last page and then uncheck include shared you can end up stuck on an empty page. See if this fixes the problem by resetting the store back to the first page:

```suggestion
        tasksStore.load({ page: 1 });
```

BillClinton replied:
resolved: [9bd19ea](https://github.com/SlateFoundation/slate-cbl/pull/710/commits/9bd19ea543cca4730c594bcee0f294406ffa6b6c)

BillClinton committed:
fix: use API method to set proxy vars

Also, reset to first page when changing filter variables to
avoid blank pages if result set shrinks
